### ISSUE TITLE
Support consistent data_format between tf.layers and everything else

### ISSUE THREAD

**Issue by taion** [tensorflow/tensorflow#8137]
The functions in `tf.layers` take a `data_format` parameter. However, this parameter has different semantics from the identically named `data_format` parameter everywhere else in TensorFlow. It's expected to be `channels_first` or `channels_last`, versus `NHWC`, `NCHW`, or `NDHWC` everywhere else. As such, it's inconvenient from a DX perspective to intersperse `tf.layers` code with other TensorFlow code, as it requires passing different values for the identically-named `data_format` parameter.

Ideally, the functions in `tf.layers` should support the more explicit `data_format` strings. While `channels_first` and `channels_last` are easier to understand, they're less explicit, as there do exist cases outside of TensorFlow where the tensor layout is CHWN, given which `channels_first` meaning `NCHW` is not optimally clear.

On the same note, it's a bit inconvenient that `tf.layers.batch_normalization` takes `axis` instead of `data_format`; while this is more correct, it makes it annoying to switch back and forth, especially that the fused batch norm implementation only supports NHWC and NCHW anyway, rather than batch norm on an arbitrary axis.

**Comment by yaroslavvb** [comment]
Drive-by comment -- `tf.layers.conv2d` doesn't support NCHW on CPU because of BiasOp. Seems weird that op is still being used, a+b should be sufficient

**Comment by taion** [comment]
@yaroslavvb I haven't checked, but per https://github.com/tensorflow/tensorflow/issues/7551, it might be slower to use the broadcasted add instead of `bias_add`.

**Comment by taion** [comment]
Yup, `tf.nn.bias_add` seems to be faster than doing a broadcasted addition: https://github.com/tensorflow/tensorflow/issues/7551#issuecomment-284519625

**Comment by martinwicke** [comment]
We could add support for the relevant strings to layers easily. The question is what to do for the modes we don't support. @fchollet FYI.

**Comment by fchollet** [comment]
It's planned to add support for `data_format` argument values such as `NHWC`, etc. It will be easy, too, and we'll get to it soon. @martinwicke: what do you mean by "modes we don't support", is it e.g. `C*` formats? Or is it the fact that a lot of TF ops support some formats but not others? (e.g. `conv3d`, `bias_add`, etc...

**Comment by martinwicke** [comment]
Exactly, what to we do when we encounter a partially unsupported format. I guess we could let the inside ops error out if they wish. When we start allowing HW and WH modes we also have to be careful about what should happen when they are mixed (can they be?)

**Comment by itsmeolivia** [comment]
Automatically closing due to lack of recent activity. Please update the issue when new information becomes available, and we will reopen the issue. Thanks!

**Comment by taion** [comment]
There isn't any new information, but this issue is still relevant. It would be really nice to not have to convert back and forth between `NCHW` and `channels_first` in switching between API levels.

**Comment by martinwicke** [comment]
@tfboyd @tatianashp if I understand correctly, this will become more pertinent with MKL, for which the default (or only supported) data format is different from GPU?

@rmlarsen data layout optimization is a grappler pass, right? 

What is the way forward to relieve actual users from these problems?

**Comment by taion** [comment]
The right answer there is something like https://github.com/tensorflow/tensorflow/issues/8227.

This is more of an smaller ergonomics thing. I just want to be able to do `data_format='NCHW'` for e.g. `tf.layers.conv2d`, as is the case for most other things in TF, rather than having to do `data_format='channels_first'`, which is the only supported scheme in `tf.layers`, and is not supported anywhere else in TF.

**Comment by fchollet** [comment]
@taion This is a solved issue. Consider using `tf.keras` to build your models. In all layers from `tf.keras.layers` that take a `data_format` argument, the default will be `keras.backend.image_data_format()` (which you can set via `keras.backend.set_image_data_format('channels_...')` or via your Keras settings files). This allows you to build models that work for either data format, without changing anything.

Agreed that we should extend support to `BatchNormalization`.

**Comment by taion** [comment]
We should probably leave discussion there on #8227.

The specific issue here is just that it's awkward to write:

```python
net = tf.layers.conv2d(net, ..., data_format='channels_first')
net = tf.whatever(net, ..., data_format='NCHW')
```

**Comment by sharvil** [comment]
@martinwicke, could you please help shepherd this issue towards resolution? It would be nice to have a consistent set of strings for `data_format` across all of TF.

**Comment by andreydung** [comment]
@yaroslavvb: I'm wondering if you have any inputs on this. Currently, is it possible to convert a Pytorch model that inputs NCHW (3, 224, 224) to a TFLite model that inputs NHWC (224, 224, 3)? I was able to convert Pytorch NCHW to TFLite NCHW but the generated model is complicated (10 times increase in memory and inference speed).
